// @ngui/map/.ng_build/ts/directives/marker.js
import { Directive } from '@angular/core';
import { BaseMapDirective } from './base-map-directive';
import { NguiMapComponent } from '../components/ngui-map.component';
const /** @type {?} */ INPUTS = [
    'anchorPoint', 'animation', 'clickable', 'cursor', 'draggable', 'icon', 'label', 'opacity',
    'optimized', 'place', 'position', 'shape', 'title', 'visible', 'zIndex', 'options',
    // ngui-map specific inputs
    'geoFallbackPosition'
];
const /** @type {?} */ OUTPUTS = [
    'animationChanged', 'click', 'clickableChanged', 'cursorChanged', 'dblclick', 'drag', 'dragend', 'draggableChanged',
    'dragstart', 'flatChanged', 'iconChanged', 'mousedown', 'mouseout', 'mouseover', 'mouseup', 'positionChanged', 'rightclick',
    'shapeChanged', 'titleChanged', 'visibleChanged', 'zindexChanged'
];
export class Marker extends BaseMapDirective {
    /**
     * @param {?} nguiMapComp
     */
    constructor(nguiMapComp) {
        super(nguiMapComp, 'Marker', INPUTS, OUTPUTS);
        this.nguiMapComp = nguiMapComp;
        this.objectOptions = ({});
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.nguiMapComponent.mapIdledOnce) {
            this.initialize();
        }
        else {
            this.nguiMapComponent.mapReady$.subscribe(map => this.initialize());
        }
    }
    /**
     * @return {?}
     */
    initialize() {
        super.initialize();
        this.setPosition();
    }
    /**
     * @return {?}
     */
    setPosition() {
        if (!this['position']) {
            this._subscriptions.push(this.nguiMapComp.geolocation.getCurrentPosition().subscribe(position => {
                let /** @type {?} */ latLng = new google.maps.LatLng(position.coords.latitude, position.coords.longitude);
                this.mapObject.setPosition(latLng);
            }, error => {
                console.error('ngui-map, error finding the current location');
                this.mapObject.setPosition(this.objectOptions['geoFallbackPosition'] || new google.maps.LatLng(0, 0));
            }));
        }
        else if (typeof this['position'] === 'string') {
            this._subscriptions.push(this.nguiMapComp.geoCoder.geocode({ address: this['position'] }).subscribe(results => {
                this.mapObject.setPosition(results[0].geometry.location);
            }, error => {
                console.error('ngui-map, error finding the location from', this['position']);
                this.mapObject.setPosition(this.objectOptions['geoFallbackPosition'] || new google.maps.LatLng(0, 0));
            }));
        }
    }
}
Marker.decorators = [
    { type: Directive, args: [{
                selector: 'ngui-map > marker',
                inputs: INPUTS,
                outputs: OUTPUTS,
            },] },
];
/**
 * @nocollapse
 */
Marker.ctorParameters = () => [
    { type: NguiMapComponent, },
];
function Marker_tsickle_Closure_declarations() {
    /** @type {?} */
    Marker.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    Marker.ctorParameters;
    /** @type {?} */
    Marker.prototype.anchorPoint;
    /** @type {?} */
    Marker.prototype.animation;
    /** @type {?} */
    Marker.prototype.clickable;
    /** @type {?} */
    Marker.prototype.cursor;
    /** @type {?} */
    Marker.prototype.draggable;
    /** @type {?} */
    Marker.prototype.icon;
    /** @type {?} */
    Marker.prototype.label;
    /** @type {?} */
    Marker.prototype.opacity;
    /** @type {?} */
    Marker.prototype.optimized;
    /** @type {?} */
    Marker.prototype.place;
    /** @type {?} */
    Marker.prototype.position;
    /** @type {?} */
    Marker.prototype.shape;
    /** @type {?} */
    Marker.prototype.title;
    /** @type {?} */
    Marker.prototype.visible;
    /** @type {?} */
    Marker.prototype.zIndex;
    /** @type {?} */
    Marker.prototype.options;
    /** @type {?} */
    Marker.prototype.geoFallbackPosition;
    /** @type {?} */
    Marker.prototype.animationChanged;
    /** @type {?} */
    Marker.prototype.click;
    /** @type {?} */
    Marker.prototype.clickableChanged;
    /** @type {?} */
    Marker.prototype.cursorChanged;
    /** @type {?} */
    Marker.prototype.dblclick;
    /** @type {?} */
    Marker.prototype.drag;
    /** @type {?} */
    Marker.prototype.dragend;
    /** @type {?} */
    Marker.prototype.draggableChanged;
    /** @type {?} */
    Marker.prototype.dragstart;
    /** @type {?} */
    Marker.prototype.flatChanged;
    /** @type {?} */
    Marker.prototype.iconChanged;
    /** @type {?} */
    Marker.prototype.mousedown;
    /** @type {?} */
    Marker.prototype.mouseout;
    /** @type {?} */
    Marker.prototype.mouseover;
    /** @type {?} */
    Marker.prototype.mouseup;
    /** @type {?} */
    Marker.prototype.positionChanged;
    /** @type {?} */
    Marker.prototype.rightclick;
    /** @type {?} */
    Marker.prototype.shapeChanged;
    /** @type {?} */
    Marker.prototype.titleChanged;
    /** @type {?} */
    Marker.prototype.visibleChanged;
    /** @type {?} */
    Marker.prototype.zindexChanged;
    /** @type {?} */
    Marker.prototype.mapObject;
    /** @type {?} */
    Marker.prototype.objectOptions;
    /** @type {?} */
    Marker.prototype.nguiMapComp;
}
