// @ngui/map/.ng_build/ts/directives/ground-overlay.js
import { Directive } from '@angular/core';
import { BaseMapDirective } from './base-map-directive';
import { NguiMapComponent } from '../components/ngui-map.component';
const /** @type {?} */ INPUTS = ['url', 'bounds', 'clickable', 'opacity'];
const /** @type {?} */ OUTPUTS = ['click', 'dblclick'];
export class GroundOverlay extends BaseMapDirective {
    /**
     * @param {?} nguiMapComp
     */
    constructor(nguiMapComp) {
        super(nguiMapComp, 'GroundOverlay', INPUTS, OUTPUTS);
        this.objectOptions = ({});
    }
    /**
     * @return {?}
     */
    initialize() {
        // url, bounds are not the options of GroundOverlay
        this.objectOptions = this.optionBuilder.googlizeAllInputs(['clickable', 'opacity'], this);
        // noinspection TypeScriptUnresolvedFunction
        this.mapObject = new google.maps.GroundOverlay(this['url'], this['bounds'], this.objectOptions);
        this.mapObject.setMap(this.nguiMapComponent.map);
        this.mapObject['mapObjectName'] = this.mapObjectName;
        // set google events listeners and emits to this outputs listeners
        this.nguiMap.setObjectEvents(this.outputs, this, 'mapObject');
        this.nguiMapComponent.addToMapObjectGroup(this.mapObjectName, this.mapObject);
        this.initialized$.emit(this.mapObject);
    }
}
GroundOverlay.decorators = [
    { type: Directive, args: [{
                selector: 'ngui-map > ground-overlay',
                inputs: INPUTS,
                outputs: OUTPUTS,
            },] },
];
/**
 * @nocollapse
 */
GroundOverlay.ctorParameters = () => [
    { type: NguiMapComponent, },
];
function GroundOverlay_tsickle_Closure_declarations() {
    /** @type {?} */
    GroundOverlay.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    GroundOverlay.ctorParameters;
    /** @type {?} */
    GroundOverlay.prototype.url;
    /** @type {?} */
    GroundOverlay.prototype.bounds;
    /** @type {?} */
    GroundOverlay.prototype.clickable;
    /** @type {?} */
    GroundOverlay.prototype.opacity;
    /** @type {?} */
    GroundOverlay.prototype.click;
    /** @type {?} */
    GroundOverlay.prototype.dblclick;
    /** @type {?} */
    GroundOverlay.prototype.mapObject;
    /** @type {?} */
    GroundOverlay.prototype.objectOptions;
}
