// @ngui/map/.ng_build/ts/directives/heatmap-layer.js
import { Directive } from '@angular/core';
import { BaseMapDirective } from './base-map-directive';
import { NguiMapComponent } from '../components/ngui-map.component';
const /** @type {?} */ INPUTS = ['data', 'dissipating', 'gradient', 'maxIntensity', 'opacity', 'radius', 'options'];
const /** @type {?} */ OUTPUTS = [];
export class HeatmapLayer extends BaseMapDirective {
    /**
     * @param {?} nguiMapComp
     */
    constructor(nguiMapComp) {
        super(nguiMapComp, 'HeatmapLayer', INPUTS, OUTPUTS);
        this.libraryName = 'visualization';
    }
}
HeatmapLayer.decorators = [
    { type: Directive, args: [{
                selector: 'ngui-map > heatmap-layer',
                inputs: INPUTS,
                outputs: OUTPUTS,
            },] },
];
/**
 * @nocollapse
 */
HeatmapLayer.ctorParameters = () => [
    { type: NguiMapComponent, },
];
function HeatmapLayer_tsickle_Closure_declarations() {
    /** @type {?} */
    HeatmapLayer.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    HeatmapLayer.ctorParameters;
    /** @type {?} */
    HeatmapLayer.prototype.data;
    /** @type {?} */
    HeatmapLayer.prototype.dissipating;
    /** @type {?} */
    HeatmapLayer.prototype.gradient;
    /** @type {?} */
    HeatmapLayer.prototype.maxIntensity;
    /** @type {?} */
    HeatmapLayer.prototype.opacity;
    /** @type {?} */
    HeatmapLayer.prototype.radius;
    /** @type {?} */
    HeatmapLayer.prototype.options;
    /** @type {?} */
    HeatmapLayer.prototype.libraryName;
}
